docs/data/data-grid/column-menu/CustomColumnMenuGrid.js

```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import {
+  DataGrid,
+  GridColumnMenuFilterItem,
+  GridColumnMenuSortItem,
+  GridColumnMenuColumnsItem,
+} from '@mui/x-data-grid';
+import { useDemoData } from '@mui/x-data-grid-generator';
+import Stack from '@mui/material/Stack';
+import Button from '@mui/material/Button';
+import Divider from '@mui/material/Divider';
+
+function MenuCloseComponent(props) {
+  return (
+    <Button color="primary" onClick={props.onClick}>
+      Close Menu
+    </Button>
+  );
+}
+
+MenuCloseComponent.propTypes = {
+  onClick: PropTypes.func.isRequired,
+};
+
+function CustomColumnMenu(props) {
+  const itemProps = {
+    column: props.currentColumn,
+    onClick: props.hideMenu,
+  };
+  return (
+    <Stack p={0.5}>
+      <GridColumnMenuSortItem {...itemProps} />
+      {/* Only provide filtering on desk */}
+      {itemProps.column.field === 'desk' ? (
+        <GridColumnMenuFilterItem {...itemProps} />
+      ) : null}
+      <Divider />
+      <GridColumnMenuColumnsItem {...itemProps} />
+      <MenuCloseComponent {...itemProps} />
```

Comment:
Is this the recommended way of adding new items to the column menu?
I wonder if there's a simpler way to add an item without rebuilding the default menu items

Reply:
I previously added the option to add custom components using `components` prop and then register them using something like:
```tsx
<GridColumnMenuDefault
  {...props}
  components={{
    SomeCustomSlot: CustomComponent
  }}
  componentsProps={{
    SomeCustomSlot: {
      displayOrder: x;
    }
  }}
  initialItems=['SomeCustomSlot'] // register user items
/>
```

But it:
1. Didn't seem consistent with how we treat `rootProps.components`
2. More props mean API is complex

What do you think about this interface though?

Reply:
For (1) and (2), we can implicitly register any custom component:
```tsx
<GridColumnMenuDefault
  {...props}
  components={{
    SomeCustomSlot: CustomComponent
  }}
/>
```

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenuComponents.ts b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenuComponents.ts
index 34308d9..50a422e 100644
--- a/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenuComponents.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/columnMenu/useGridColumnMenuComponents.ts
@@ -44,6 +44,10 @@ const useGridColumnMenuComponents = (
     const sorted = preProcessedItems.sort(
       (a, b) => processedComponentsProps[a].displayOrder - processedComponentsProps[b].displayOrder,
     );
+    const customComponents = Object.keys(processedComponents).filter(
+      (key) => !preProcessedItems.includes(key),
+    );
+    sorted.push(...customComponents);
     // Future Enhancement, pass other `componentsProps` to respective components if needed
     return sorted.reduce<React.JSXElementConstructor<any>[]>((acc, key) => {
       if (!processedComponents[key]) {
```

For (3), we can push items without `displayOrder` to the end of the menu.

Also, to make menu items added like this look good by default, maybe apply padding one level higher, here?

https://github.com/MBilalShafi/mui-x/blob/38e475482a8a11e1beb63737ba17e5520babfefe/packages/grid/x-data-grid/src/components/menu/columnMenu/default/GridColumnMenuDefault.tsx#L49

Reply:
> For (1) and (2), we can implicitly register any custom component

Yes, it makes sense, I've done it.

> For (3), we can push items without displayOrder to the end of the menu.

Yes, I've set a default displayOrder i.e. `100` to push the items by default to the end of list

> Also, to make menu items added like this look good by default, maybe apply padding one level higher, here?

Shouldn't it be better to leave this up to the user's discretion, as they may sometimes don't need padding or want to customize it as per their needs?

Reply:
> Shouldn't it be better to leave this up to the user's discretion, as they may sometimes don't need padding or want to customize it as per their needs?

We should definitely allow users to customize it.
But when adding items to the default menu, it should look more or less consistent with the rest of the items.